pynitrokey: 0.4.9 -> 0.4.26

sbruder opened the pull request:
###### Description of changes

This updates pynitrokey as part of the [Summer of Nix](https://summer.nixos.org/).
It introduces a few new packages that are new dependencies of pynitrokey.

A few notes on how this is packaged:

 * It depends on https://github.com/NixOS/nixpkgs/pull/180075, which fixes the nrfutil package, a new dependency of pynitrokey. This PR’s branch includes the commits from that PR to make it build self-contained, but it should not be merged before the other PR is merged.
 * It does not fully use pypemicro, which is an optional dependency of some projects. Where it is a hard dependency, it is patched out. The reason is that it packages prebuilt shared libraries for which no source code is available: https://github.com/NXPmicro/pypemicro/issues/10
 * Since I can’t guarantee that I will be able to maintain this after Summer of Nix, I also added @frogamic as a maintainer to each package, I hope this is okay for you.
 * I don’t own a Nitrokey device and couldn’t test this with real hardware. All (well, most, cf. spsdk) tests for the dependencies pass and the CLI looks functional, so I think it should work. @devhell @frogamic, you were involved in initially adding the package, could you please test if this works with a Nitrokey?
 * I also ran into https://github.com/NixOS/nixpkgs/issues/176917, which I addressed by adding `binutils.bintools` to the downstream packages and wrapping the binaries with a `LD_LIBRARY_PATH` including OpenSSL. I don’t know if this is the preferred way to do this.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

sbruder committed:
python3Packages.argparse_addons: init at 0.8.0

sbruder committed:
python3Packages.bincopy: init at 17.10.2

sbruder committed:
libusbsio: init at 2.1.11

Many thanks to David Runge (dvzrv), who made an effort to get this
available under a free license earlier this year:
https://github.com/NXPmicro/spsdk/issues/36

sbruder committed:
python3Packages.libusbsio: init

Its version is synced with libusbsio.

sbruder committed:
python3Packages.cmsis-pack-manager: init at 0.4.0

sbruder committed:
python3Packages.hexdump: init at 3.3

sbruder committed:
pyocd: init at 0.34.1

sbruder committed:
python3Packages.spsdk: init at 1.6.3

sbruder committed:
pynitrokey: 0.4.9 -> 0.4.26

risicle commented on pkgs/development/libraries/pc-ble-driver/default.nix:
```diff
@@ -15,6 +15,14 @@ stdenv.mkDerivation rec {
+  patches = [
+     # Fix build with GCC 11
+    (fetchpatch {
+      url = "https://github.com/NordicSemiconductor/pc-ble-driver/commit/37258e65bdbcd0b4369ae448faf650dd181816ec.patch";
```
```suggestion
      name = "gcc11-fix.patch";
      url = "https://github.com/NordicSemiconductor/pc-ble-driver/commit/37258e65bdbcd0b4369ae448faf650dd181816ec.patch";
```

risicle commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,38 @@
+  };
+
+  postPatch = ''
+    rm -r bin/*
```
:+1: great habit

risicle commented on pkgs/development/embedded/pyocd/default.nix:
```diff
@@ -0,0 +1,65 @@
+  patches = [
+    # Remove requirement for libusb-package (https://github.com/pyocd/pyOCD/pull/1332)
+    (fetchpatch {
+      url = "https://github.com/pyocd/pyOCD/commit/0b980cf253e3714dd2eaf0bddeb7172d14089649.patch";
```
```suggestion
      name = "libusb-package-optional.patch";
      url = "https://github.com/pyocd/pyOCD/commit/0b980cf253e3714dd2eaf0bddeb7172d14089649.patch";
```

risicle commented on pkgs/development/python-modules/spsdk/default.nix:
```diff
@@ -0,0 +1,123 @@
+  patches = [
+    # Add compatibility with cryptography 37 (https://github.com/NXPmicro/spsdk/pull/43)
+    (fetchpatch {
+      url = "https://github.com/NXPmicro/spsdk/commit/a85b854de1093de593d27fa64de442224ab2e0fd.patch";
```
```suggestion
      name = "cryptography-37-compat.patch";
      url = "https://github.com/NXPmicro/spsdk/commit/a85b854de1093de593d27fa64de442224ab2e0fd.patch";
```

risicle replied:
I think you get the idea and can come up with your own patch names. Often a good patch name can satisfy the need for a comment.

risicle commented on pkgs/development/embedded/pyocd/default.nix:
```diff
@@ -0,0 +1,65 @@
+    pyusb
+    pyyaml
+    typing-extensions
+  ];
```
```suggestion
  ] ++ lib.optionals (!stdenv.isLinux) [
    hidapi
  ];
```

...and obviously also add it to the args at the top.

risicle commented on pkgs/development/python-modules/cmsis-pack-manager/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+    sourceRoot = name + "/rust";
+    cargoLock.lockFile = ./Cargo.lock;
+
```
This needs `darwin.apple_sdk.frameworks.Security` added to `buildInputs` here on darwin.

risicle commented:
Looking good. I'd suggest adding `pythonImportsCheck` to all the python packages, even those with tests enabled, as it can sometimes catch problems the tests miss.

gytis-ivaskevicius commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,38 @@
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [
+    libusb
+    systemd.dev # libudev
```
`systemdMinimal.dev` should work as well and its smaller

gytis-ivaskevicius commented on pkgs/development/python-modules/spsdk/default.nix:
```diff
@@ -0,0 +1,128 @@
+        --replace "pylink-square>=0.8.2,<0.11.9" "pylink-square" \
+        --replace "pyocd-pemicro>=1.1.1,<1.2.0" "" \
+        --replace "pyocd>=0.28.3,<=0.31.0" "pyocd" \
+        --replace "pypemicro>=0.1.9,<0.2.0" ""
```
Consider packaging this dependency or writing up a comment on why is it disabled

sbruder committed:
python3Packages.pypemicro: init at 0.1.9

DamienCassou commented:
I rebased the PR on top of latest master and will merge it in the coming days if nobody says otherwise (or merges before me).

DamienCassou commented:
Works fine with my Nitrokey start.

SuperSandro2000 commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,38 @@
+{ lib, stdenv, fetchzip, pkg-config, libusb, systemdMinimal, ... }:
+let
+  libSuffix = if stdenv.isDarwin then "dylib" else "so";
```
Use ``stdenv.hostPlatform.extensions.sharedLibrary`` instead.

SuperSandro2000 commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,38 @@
+  libSuffix = if stdenv.isDarwin then "dylib" else "so";
+  binDirPrefix = if stdenv.isDarwin then "osx_" else "linux_";
+in
+stdenv.mkDerivation {
```
```suggestion
stdenv.mkDerivation rec {
```

SuperSandro2000 commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,38 @@
+  version = "2.1.11";
+
+  src = fetchzip {
+    url = "https://www.nxp.com/downloads/en/libraries/libusbsio-2.1.11-src.zip";
```
```suggestion
    url = "https://www.nxp.com/downloads/en/libraries/libusbsio-${version}-src.zip";
```

SuperSandro2000 commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,38 @@
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [
+    libusb
+    systemdMinimal.dev # libudev
```
```suggestion
    systemdMinimal # libudev
```
that should be choosen automatically

SuperSandro2000 commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,38 @@
+
+  installPhase = ''
+    runHook preInstall
+    install -D bin/${binDirPrefix}$(uname -m)/libusbsio.${libSuffix} $out/lib/libusbsio.${libSuffix}
```
I don't think we want to use uname -m here. Can't we get that from a nix variable or use a wildcard?

SuperSandro2000 replied:
I don't know but I would suspect that it could be a source for impurity.

SuperSandro2000 commented on pkgs/development/python-modules/argparse_addons/default.nix:
```diff
@@ -0,0 +1,20 @@
+{ lib, buildPythonPackage, fetchPypi }:
+
+buildPythonPackage rec {
+  pname = "argparse_addons";
+  version = "0.8.0";
+
+  src = fetchPypi {
+    inherit pname version;
```
```suggestion
  pname = "argparse-addons";
  version = "0.8.0";

  src = fetchPypi {
    pname = "argparse_addons";
    inherit version;
```

SuperSandro2000 commented on pkgs/development/python-modules/argparse_addons/default.nix:
```diff
@@ -0,0 +1,20 @@
+    sha256 = "sha256-uwiBB5RNM56NLnCnYwXd41FUTixb3rrxwttWrS5tzeg=";
+  };
+
+  pythonImportsCheck = [ "argparse_addons" ];
```
Are we running the tests https://github.com/eerimoq/argparse_addons/tree/master/tests ?

SuperSandro2000 commented on pkgs/development/python-modules/cmsis-pack-manager/default.nix:
```diff
@@ -0,0 +1,76 @@
+let
+  pname = "cmsis-pack-manager";
+  version = "0.4.0";
+  name = pname + "-" + version;
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/python-modules/cmsis-pack-manager/default.nix:
```diff
@@ -0,0 +1,76 @@
+  };
+
+  native = rustPlatform.buildRustPackage {
+    name = name + "-native";
```
```suggestion
    name = "${pname}-${version}-native";
```

SuperSandro2000 commented on pkgs/development/python-modules/cmsis-pack-manager/default.nix:
```diff
@@ -0,0 +1,76 @@
+    inherit src;
+
+    buildInputs = lib.optionals stdenv.isDarwin [
+      darwin.apple_sdk.frameworks.Security
```
```suggestion
      Security
```

SuperSandro2000 commented on pkgs/development/python-modules/cmsis-pack-manager/default.nix:
```diff
@@ -0,0 +1,76 @@
+      darwin.apple_sdk.frameworks.Security
+    ];
+
+    sourceRoot = name + "/rust";
```
```suggestion
    sourceRoot = "${name}-${version}/rust";
```

SuperSandro2000 commented on pkgs/development/python-modules/cmsis-pack-manager/default.nix:
```diff
@@ -0,0 +1,76 @@
+
+    cargoBuildFlags = [ "--lib" ];
+  };
+
+  # setup.py has a hard requirement on running cargo build as part of the python build process.
+  # This just makes all cargo calls succeed, since it is built with Nix
+  fakeCargo = writeShellScriptBin "cargo" "";
```
Can't we just patch that out?

SuperSandro2000 replied:
Can't we tell milksnake to run true as a build command?

SuperSandro2000 commented on pkgs/development/python-modules/cmsis-pack-manager/default.nix:
```diff
@@ -0,0 +1,76 @@
+
+  preBuild = ''
+    mkdir -p rust/target/release/deps
+    ln -s ${native}/lib/libcmsis_cffi.* rust/target/release/deps/
```
I doubt that the wildcard will work with multiple files.

SuperSandro2000 commented on pkgs/development/python-modules/cmsis-pack-manager/default.nix:
```diff
@@ -0,0 +1,76 @@
+
+  preCheck = ''
+    # Otherwise the test uses a dummy library (missing all symbols)
+    ln -sf ../build/lib/cmsis_pack_manager/_native__lib.so cmsis_pack_manager/_native__lib.so
```
That won't work on darwin.

SuperSandro2000 commented on pkgs/development/python-modules/hexdump/default.nix:
```diff
@@ -0,0 +1,32 @@
+    extension = "zip";
+  };
+
+  # the source zip has no prefix, so everything gets unpacked to /build otherwise
```
Can we workaround this with ``stripRoot = false``?

SuperSandro2000 commented on pkgs/development/python-modules/libusbsio/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+buildPythonPackage rec {
+  pname = "libusbsio";
+  version = libusbsio.version;
```
```suggestion
  inherit (libusbsio) version;
```

SuperSandro2000 commented on pkgs/development/python-modules/pypemicro/default.nix:
```diff
@@ -0,0 +1,20 @@
+  };
+
+  pythonImportsCheck = [ "pypemicro" ];
+
```
Are we executing the tests https://github.com/NXPmicro/pypemicro/tree/master/tests ?

SuperSandro2000 replied:
Then please add ``doCheck = false`` with a short comment

SuperSandro2000 commented on pkgs/development/python-modules/spsdk/default.nix:
```diff
@@ -0,0 +1,123 @@
+  pname = "spsdk";
+  version = "1.6.3";
+
+  # pypi does not include tests (https://github.com/NXPmicro/spsdk/issues/37)
```
```suggestion
```

SuperSandro2000 replied:
That is so standard in python packages that it isn't worth a comment

SuperSandro2000 commented on pkgs/development/python-modules/spsdk/default.nix:
```diff
@@ -0,0 +1,123 @@
+  ];
+
+  postPatch = ''
+    substituteInPlace requirements.txt \
+        --replace "cmsis-pack-manager<0.3.0" "cmsis-pack-manager" \
+        --replace "cryptography>=3.4.4,<37" "cryptography" \
+        --replace "deepmerge==0.3.0" "deepmerge" \
+        --replace "jinja2>=2.11,<3.1" "jinja2" \
+        --replace "pylink-square>=0.8.2,<0.11.9" "pylink-square" \
+        --replace "pyocd-pemicro>=1.1.1,<1.2.0" "" \
+        --replace "pyocd>=0.28.3,<=0.31.0" "pyocd"
```
Can we use the relax hook instead?

SuperSandro2000 commented on pkgs/development/python-modules/spsdk/default.nix:
```diff
@@ -0,0 +1,123 @@
+    asn1crypto
+    astunparse
+    bincopy
+    binutils.bintools # required for ctypes to find libcrypto (https://github.com/dbt-labs/dbt-core/issues/3366#issuecomment-849764321)
```
Can we patch the code instead?

SuperSandro2000 replied:
thats strange

FRidh replied:
binutils (objdump) is indeed one of the methods for ctypes to find dependencies.

Still, in Nixpkgs we patch out ctypes calls in the calling libraries. This is not for the interpreter.

SuperSandro2000 commented on pkgs/development/python-modules/spsdk/default.nix:
```diff
@@ -0,0 +1,123 @@
+    voluptuous
+  ];
+
+  # They also fail when running them pip-installed on a debian system.
+  disabledTests = [
```
```suggestion
  disabledTests = [
    # tests also fail on debian, so presumable they are broken
```

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -649,6 +649,8 @@ in {
 
   argon2-cffi-bindings = callPackage ../development/python-modules/argon2-cffi-bindings { };
 
+  argparse_addons = callPackage ../development/python-modules/argparse_addons { };
```
```suggestion
  argparse-addons = callPackage ../development/python-modules/argparse-addons { };
```

sbruder committed:
python3Packages.oscrypto: patch ctypes call

The ctypes call did not work, because it requires binutils.bintools to
be available in every downstream dependency.

Fixes #176917.

SuperSandro2000 commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,37 @@
+{ lib, stdenv, fetchzip, pkg-config, libusb, systemdMinimal, ... }:
```
```suggestion
{ lib, stdenv, fetchzip, pkg-config, libusb1, systemdMinimal }:
```

SuperSandro2000 commented on pkgs/development/libraries/libusbsio/default.nix:
```diff
@@ -0,0 +1,37 @@
+
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [
+    libusb
```
```suggestion
    libusb1
```
